torch/nativert/graph/Graph.h

```diff
@@ -0,0 +1,696 @@
+#pragma once
+
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <variant>
+#include <vector>
+
+#include <ATen/core/ivalue.h>
+#include <c10/util/IntrusiveList.h>
+#include <c10/util/Logging.h>
+
+#include <torch/csrc/utils/generated_serialization_types.h>
+#include <torch/nativert/executor/Placement.h>
+#include <torch/nativert/graph/GraphSignature.h>
+#include <torch/nativert/graph/TensorMeta.h>
+
+namespace torch::nativert {
+
+using NodeIndex = size_t;
+
+class Value;
+
+class Type {
+ public:
+  enum Kind {
+    None,
+    Tensor,
+    TensorList,
+    OptionalTensorList,
+    SymInt,
+    SymIntList,
+    SymBool,
+    SymFloat,
+    CustomObj,
+  };
+
+  friend std::ostream& operator<<(std::ostream& out, const Type& ty);
+
+  /* implicit */ Type(Kind kind) : kind_(kind) {}
+
+  explicit Type(Kind kind, const std::string& classFqn)
+      : kind_(kind), classFqn_(classFqn) {
+    CHECK(kind == Kind::CustomObj);
+    CHECK(!classFqn.empty());
+  }
+
+  Kind kind() const {
+    return kind_;
+  }
+
+  std::string classFqn() const {
+    CHECK(kind_ == Kind::CustomObj)
+        << "Only CustomObj type can have classFqn, got " << kind_;
+    return classFqn_;
+  }
```

Comment:
why waste the memory for it for all other kinds then? instead, you could represent this class with std::variant internally and materialize the Kind on access